File: backend/src/entities/Prediction.ts
import "reflect-metadata";
import { Entity, PrimaryGeneratedColumn, Column, CreateDateColumn, ManyToOne, JoinColumn } from "typeorm";
import { User } from "./User";

@Entity("predictions")
export class Prediction {
  @PrimaryGeneratedColumn("uuid")
  id: string;

  @Column({ type: "uuid" })
  userId: string;

  @ManyToOne(() => User, user => user.predictions)
  @JoinColumn({ name: "userId" })
  user: User;

  // Personal Details
  @Column({ type: "varchar", length: 255, nullable: true })
  patientName: string;

  @Column({ type: "int" })
  age: number;

  @Column({ type: "varchar", length: 20, nullable: true })
  gender: string; // Male, Female, Other

  @Column({ type: "int", nullable: true })
  pregnancies: number; // Only for females

  // Medical Measurements
  @Column({ type: "int", nullable: true })
  systolicBP: number;

  @Column({ type: "int", nullable: true })
  diastolicBP: number;

  @Column({ type: "decimal", precision: 10, scale: 2 })
  bmi: number;

  @Column({ type: "decimal", precision: 10, scale: 2, nullable: true })
  hba1c: number;

  @Column({ type: "decimal", precision: 10, scale: 2, nullable: true })
  fastingGlucose: number;

  @Column({ type: "boolean", default: false })
  familyHistory: boolean;

  // Legacy fields (kept for backward compatibility)
  @Column({ type: "int", nullable: true })
  diastolicBloodPressure: number;

  @Column({ type: "decimal", precision: 10, scale: 2, nullable: true })
  serumInsulin: number;

  @Column({ type: "int", nullable: true })
  skinFoldThickness: number;

  // Lifestyle Factors
  @Column({ type: "varchar", length: 50, nullable: true })
  physicalActivity: string; // No Activity, Little Activity, Moderate Activity, High Activity

  @Column({ type: "boolean", nullable: true })
  smoking: boolean;

  @Column({ type: "boolean", nullable: true })
  alcohol: boolean;

  // Symptoms
  @Column({ type: "boolean", nullable: true })
  excessiveThirst: boolean;

  @Column({ type: "boolean", nullable: true })
  frequentUrination: boolean;

  @Column({ type: "boolean", nullable: true })
  suddenWeightLoss: boolean;

  // Results
  @Column({ type: "varchar", length: 50 })
  riskLevel: string; // Low, Moderate, High

  @Column({ type: "decimal", precision: 5, scale: 4 })
  probability: number;

  @CreateDateColumn()
  createdAt: Date;
}

File: backend-python/models/prediction.py
from sqlalchemy import Column, String, Integer, Boolean, DateTime, Numeric, ForeignKey
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import relationship
from database.database import Base
import uuid

class Prediction(Base):
    __tablename__ = "predictions"
    
    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    user_id = Column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=False)
    
    # Personal Details
    patient_name = Column(String(255), nullable=True)
    age = Column(Integer, nullable=False)
    gender = Column(String(20), nullable=True)  # Male, Female, Other
    pregnancies = Column(Integer, nullable=True)  # Only for females
    
    # Medical Measurements
    systolic_bp = Column(Integer, nullable=True)
    diastolic_bp = Column(Integer, nullable=True)
    bmi = Column(Numeric(10, 2), nullable=False)
    hba1c = Column(Numeric(10, 2), nullable=True)
    fasting_glucose = Column(Numeric(10, 2), nullable=True)
    family_history = Column(Boolean, default=False)
    
    # Legacy fields (kept for backward compatibility)
    diastolic_blood_pressure = Column(Integer, nullable=True)
    serum_insulin = Column(Numeric(10, 2), nullable=True)
    skin_fold_thickness = Column(Integer, nullable=True)
    
    # Lifestyle Factors
    physical_activity = Column(String(50), nullable=True)  # No Activity, Little Activity, Moderate Activity, High Activity
    smoking = Column(Boolean, nullable=True)
    alcohol = Column(Boolean, nullable=True)
    
    # Symptoms
    excessive_thirst = Column(Boolean, nullable=True)
    frequent_urination = Column(Boolean, nullable=True)
    sudden_weight_loss = Column(Boolean, nullable=True)
    
    # Results
    risk_level = Column(String(50), nullable=False)  # Low, Moderate, High
    probability = Column(Numeric(5, 4), nullable=False)
    
    created_at = Column(DateTime, nullable=False)
    
    # Relationship
    user = relationship("User", back_populates="predictions")
